pkg/pss/pss.go

```diff
@@ -116,18 +116,27 @@ func (p *pss) Register(topic trojan.Topic, handler Handler) (cleanup func()) {
 	}
 }
 
+func (p *pss) topics() []Topic {
+	p.handlersMu.Lock()
+	defer p.handlersMu.Unlock()
+
+	ts := make([]Topic, 0, len(p.handlers))
+	for t := range p.handlers {
```

Comment:
this loop shouldnt use append then

Reply:
I think comment @janos made was on the point. The pre-allocated array creates slice with length 0 (but enough capacity). Append operations should add into existing slice until capacity is reached, after which each append will reallocate and create new slice. Since we only ever add up to `len(p.handlers)` we should be fine.

Reply:
i agree with viktor about this. as a convention using `make` then append is possible source of trouble and we've had some pretty serious bugs due to this. it would also prefer this not to use append

Reply:
Sorry for the confusion, I thought that it is an obvious thing. Feel free to remove it.

Reply:
I personally do not have special preference to specific way (although I prefer more optimized code).

The options for this are now following (omiting mutex):

- named return value

```
func (p *pss) topics() (ts []Topic) {
	for t, _ := range p.handlers {
		ts = append(ts, t)
	}
	return ts
}
```

- exactly allocated slice with append

```
func (p *pss) topics() []Topic {
	ts := make([]Topic, 0, len(p.handlers))
	for t := range p.handlers {
		ts = append(ts, t)
	}
	return ts
}
```

- exactly allocated slice with helper variable for index

```
func (p *pss) topics() []Topic {
	ts := make([]Topic, len(p.handlers))
    i := 0
	for t := range p.handlers {
		ts[i] = t
        i++
	}
	return ts
}
```

If someone has some other suggestion please add so it can be decided.

There is no special reason to select the code with additional variable, but if that looks easier, we can change to that one.

Reply:
no need to explicitly declare the helper variable as it is provided by `range`:
```
func (p *pss) topics() []Topic {
	ts := make([]Topic, len(p.handlers))
	for i, t := range p.handlers {
		ts[i] = t
	}
	return ts
}
```

Reply:
It is needed in this case, as the range is over the map, it does not have the index.

Reply:
@acud - "as a convention using make then append is possible source of trouble and we've had some pretty serious bugs due to this" - could you elaborate more pls? I am interested in this convention.

Reply:
> It is needed in this case, as the range is over the map, it does not have the index.

true sorry.

Reply:
> could you elaborate more pls

as an example we had a problem in old swarm codebase where we had errors when setting chunks as synced in localstore.
this was caused by the fact that we used:
```go
someSlice := make([]chunkType, expectedSize)
<then later...>
someSlice = append(someSlice, chunk)
```
So obviously the first entries were empty, and the context down the line did not try to fulfill all elements in the slice, but returned an error on the first item that errored (the first one, since it was nil), as you would normally do.

It took a lot of time to find this problem and it is a very fine difference to spot out when you're skimming down a codebase, i.e. `make([]type,len,cap); then append` vs `make([]type,len); then append`, at least my opinion.

Reply:
I see, thanks for the explanation :) I still think this is more the case of "once bitten twice shy" then convention ;)

Reply:
yeah true. i meant it in the way that if we get in the habit of using this as convention we might be possibly burying some bugs under the hood. so i mean that i find the convention of `make then append` risky to use in a sense. anyway, won't block on this. total nitpick :)